core/src/es/adapter/transformer/strategy/mod.rs

```diff
@@ -0,0 +1,117 @@
+
+use crate::es::{adapter, event};
+
+use super::{Adapt, Transformer};
+
+#[doc(inline)]
+pub use self::{
+    as_is::AsIs,
+    custom::{Custom, Customize},
+    into::Into,
+    skip::Skip,
+    split::{Split, Splitter},
+};
+
+/// Generalized [`Transformer`] for [`Versioned`] events.
+///
+/// [`Versioned`]: event::Versioned
+pub trait Strategy<Adapter, Event> {
+    /// Context of this [`Strategy`].
+    ///
+    /// In real world this is usually `dyn Trait`. In that case,
+    /// [`Adapter::transform_all()`][1] will expect type which can be
+    /// [`Borrow`]ed as `dyn Trait`.
+    ///
+    /// [1]: adapter::Adapter::transform_all
+    type Context: ?Sized;
+
+    /// Error of this [`Strategy`].
+    type Error;
+
+    /// Converted [`Event`].
+    ///
+    /// [`Event`]: crate::es::Event
+    type Transformed;
+
+    /// [`Stream`] of [`Transformed`] [`Event`]s.
+    ///
+    /// [`Event`]: crate::es::Event
+    /// [`Transformed`]: Self::Transformed
+    type TransformedStream<'out>: Stream<
+            Item = Result<
+                <Self as Strategy<Adapter, Event>>::Transformed,
+                <Self as Strategy<Adapter, Event>>::Error,
+            >,
+        > + 'out;
+
+    /// Converts incoming [`Event`] into [`Transformed`].
+    ///
+    /// [`Event`]: crate::es::Event
+    /// [`Transformed`]: Self::Transformed
+    fn transform<'me: 'out, 'ctx: 'out, 'out>(
+        adapter: &'me Adapter,
+        event: Event,
+        context: &'ctx Self::Context,
+    ) -> Self::TransformedStream<'out>;
+}
+
+impl<'ctx, Event, Adapter, Ctx> Transformer<'ctx, Event, Ctx>
+    for adapter::Wrapper<Adapter>
+where
+    Event: event::VersionedOrRaw,
+    Adapter: Adapt<Event> + adapter::Returning,
+    Adapter::Strategy: Strategy<Adapter, Event>,
+    Adapter::Transformed:
+        From<<Adapter::Strategy as Strategy<Adapter, Event>>::Transformed>,
+    Adapter::Error:
+        From<<Adapter::Strategy as Strategy<Adapter, Event>>::Error>,
+    Ctx: Borrow<<Adapter::Strategy as Strategy<Adapter, Event>>::Context>
+        + ?Sized,
+    <Adapter::Strategy as Strategy<Adapter, Event>>::Context: 'ctx,
+{
+    type Error = <Adapter::Strategy as Strategy<Adapter, Event>>::Error;
+    type Transformed =
+        <Adapter::Strategy as Strategy<Adapter, Event>>::Transformed;
+    type TransformedStream<'out> = <Adapter::Strategy as Strategy<
+        Adapter,
+        Event,
+    >>::TransformedStream<'out>;
+
+    fn transform<'me, 'out>(
+        &'me self,
+        event: Event,
+        context: &'ctx Ctx,
+    ) -> Self::TransformedStream<'out>
+    where
+        'me: 'out,
+        'ctx: 'out,
+    {
+        <Adapter::Strategy as Strategy<Adapter, Event>>::transform(
+            &self.0,
+            event,
+            context.borrow(),
+        )
+    }
+}
+
+/// [`Strategy::Context`] implemented for every type.
+pub trait AnyContext {}
+
+impl<T: ?Sized> AnyContext for T {}
+
+impl Borrow<(dyn AnyContext + 'static)> for () {
```

Comment:
This part is quite akward. As far as I see, we use this `dyn AnyContext` only in places where we don't really require any real context. We pay for dynamic dispatch out of nothing.

What stops us from using `()` or some custom `EmptyContext`?

Reply:
Many issues here goes from coherence problem, thus having `impl<T: ?Sized> Borrow<T> for T` in `std` really gives us headache here.

BUT!!!1

We do really have here the situation where our code is between user's ones, so we do have control on both ends. This allows us to introduce custom wrapper types to overcome coherence rules in a way we need, use those wrappers in our inner machinery and trait bounds, and unwrap on user's ends, so library users don't ever touch them! (We do similar trick with our `event::Initial` wrapper)

Let's see...
```rust
#[derive(Clone, Copy, Debug, RefCast)]
#[repr(transparent)]
pub struct Context<T: ?Sized>(T);

#[derive(Clone, Copy, Debug, RefCast)]
#[repr(transparent)]
pub struct InnerContext<T: ?Sized>(T);

impl<T: ?Sized> Borrow<()> for Context<T> {
    fn borrow(&self) -> &() {
        &()
    }
}

impl<X: ?Sized, Y: ?Sized + Borrow<X>> Borrow<InnerContext<X>> for Context<Y> {
    fn borrow(&self) -> &Inner<X> {
        RefCast::ref_cast(self.0.borrow())
    }
}
```

Now we have `impl`s working in the way we need, and they don't impose any coherence problems.

The last part we need is to chage bounds like:
```rust
Ctx: Borrow<Ctx2>
```
to
```rust
Context<Ctx>: Borrow<Ctx2>
```
on one side, and correct our `Strategy`s `impl`s to contain either `type Context = ()` or `type Context = InnerContex<Inner::Context>`.

And, also, wrapping/unwrapping on both ends.

Let's give this idea a shot!

Reply:
@tyranron `Context<Ctx>: Borrow<Ctx2>` isn't really possible because of compiler error: ``` `?Trait` bounds are only permitted at the point where a type parameter is declared```

Reply:
@tyranron 

I've managed to work around this by wrapping `Ctx` inside `Context<Ctx>` in `Adapter` blanket impl and later working with it like `Ctx`. Default `Strategies` work just fine with it.

But there is another problem: `Custom` `Strategy`. So there's a tradeoff:

1. Require users to wrap `dyn Trait` in `InnerContext` and leave it as is in case of `()`.
2. Lift that requirement and leave them `dyn AnyContext`-like approach. 

This tradeoff arises, because for option 1 we need to provide additional support on our side, that will break option 2.

Reply:
@ilslv 

> `Context<Ctx>: Borrow<Ctx2>` isn't really possible because of compiler error: ``` `?Trait` bounds are only permitted at the point where a type parameter is declared```. 

Yeah, that's why I was (and am) insisting on using associative types. So we have the concrete type to check there from the downstream traits.

> But there is another problem: `Custom` `Strategy`. So there's a tradeoff:
>
> 1. Require users to wrap `dyn Trait` in `InnerContext` and leave it as is in case of `()`.
> 2. Lift that requirement and leave them `dyn AnyContext`-like approach. 
>
> This tradeoff arises, because for option 1 we need to provide additional support on our side, that will break option 2. 

As I've said before, it shouldn't be the case, because we can omit making users writing `InnerContext` in their signatures, by just using them in bounds.

Let's discuss it in a talk.

Reply:
@ilslv so OK, let's just use in customize either `type Context = ();` or `type Context = Borrowed<Something>`. Where the `Borrowed` is just renamed `InnerContext`.

Also, move `Borrowed` to a top-level `spell` module, so it can be imported as `use arcana::spell::Borrowed;`.

I've though quite a while about this and haven't come up with a better solution. We always may adjust it in future once the one appears.